Add create_for_localhost_exception parameter to mongodb_user module

cognifloyd opened the pull request:
##### SUMMARY
Simplify touching a file after successfully creating the first admin user under the localhost exception.

Under the mongodb_mongos molecule playbooks, extra tasks handle checking for and touching the file

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
mongodb_user module

##### ADDITIONAL INFORMATION
Simplify touching a file when creating an admin user with the localhost
exception. A pain point with the localhost exception is that you can't
tell if the first admin account has been added without attempting to
authenticate. So, provide a convenient helper that will touch the given
path when the user is successfully added.
    
Note that the name is long and unwieldy on purpose. This should be a
relatively uncommon use-case. When adding or updating most users, So, we
use the length as a signal to users that "this is not the parameter you
were looking for".

cognifloyd commented:
I'm not sure on the naming of this parameter. What do you think of doing something like this?

rhysmeister commented:
I use this hack with additional tasks to do basically the same thing. If I remember correctly they have really tied down the localhost exception and I'm not sure there's a better alternative.

cognifloyd commented:
I got the hack from your playbooks :)
Since this is something that will be useful whenever setting up a mongo server (everyone really should enable authentication in their mongo installs by default!), I think it makes sense to add this to the mongodb_user module instead of proliferating those extra tasks in many different playbooks.

So, I think this needs some kind of extra tests before merging.

In principal are you ok with putting this hack in this module?

rhysmeister commented:
Think this needs a rebase.

cognifloyd commented:
rebased. I'll look at adding integration tests next.

cognifloyd commented on plugins/modules/mongodb_user.py:
```diff
@@ -355,6 +370,10 @@ def main():
             if db_name != "admin":
                 module.fail_json(msg='The localhost login exception only allows the first admin account to be created')
             # else: this has to be the first admin user added
+            use_localhost_exception = True
+        else:
+            # Is anyone really using pymongo older than 3.0?
+            use_localhost_exception = True
```
Once #179 is merged, this will also need to make sure that `auth_mechanism` is not defined.

rhysmeister replied:
Mmm, I'm not so sure about that. From what I gather on https://github.com/ansible-collections/community.mongodb/issues/177 it would be needed for that use case. Perhaps others too?

cognifloyd replied:
Yeah. I dropped this and went back to only checking if login_user is None.

rhysmeister commented:
Looks like one of the roles is also failing due to this...

```
"/tmp/ansible_community.mongodb.mongodb_user_payload_WL6ChR/ansible_community.mongodb.mongodb_user_payload.zip/ansible_collections/community/mongodb/plugins/modules/mongodb_user.py", line 225, in user_add
    AttributeError: 'OperationFailure' object has no attribute 'keys'
    
    fatal: [config1]: FAILED! => {
        "changed": false,
        "invocation": {
            "module_args": {
                "create_for_localhost_exception": null,
                "database": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
                "login_database": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
                "login_host": "localhost",
                "login_password": null,
                "login_port": 27019,
                "login_user": null,
                "name": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
                "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
                "replica_set": null,
                "roles": [
                    "root"
                ],
                "ssl": false,
                "ssl_ca_certs": null,
                "ssl_cert_reqs": "CERT_REQUIRED",
                "ssl_certfile": null,
                "ssl_crlfile": null,
                "ssl_keyfile": null,
                "ssl_pem_passphrase": null,
                "state": "present",
                "update_password": "always"
            }
        },
        "msg": "Unable to add or update user: 'OperationFailure' object has no attribute 'keys'"
    }
```

See: https://travis-ci.com/github/ansible-collections/community.mongodb/jobs/388955210

Although I think that's probably now handled

rhysmeister commented on plugins/modules/mongodb_user.py:
```diff
@@ -202,23 +210,22 @@ def user_find(client, user, db_name):
+        # when auth is enabled on a new instance. The loalhost exception should
+        # allow us to create the first user.
+        # Might be able to do something with db.getUsers() if this approach needs a rethink.
+        if use_localhost_exception and 'code' in excep.keys() and excep.code == 13:  # 13=Unauthorized
```
This need to be something like...

if hasattr(excep, 'code') and excep.code == 13:

Tests are failing because there's no keys method on this object. I'm not really sure about the flow of how this use_localhost_exception is working.

cognifloyd replied:
> Tests are failing because there's no keys method on this object.

doh. True.

> I'm not really sure about the flow of how this use_localhost_exception is working.

I want to explicitly handle the localhost_exception, instead of assuming that the it applies and swallowing all unauthorized errors here. So, when we expect to use the localhost exception to add the first admin, then we ignore the unauthorized exception as searches should not work in this case. But, when we do not expect to use the localhost exception, then searches should work with the provided login credentials and so unauthorized errors should be evident.

rhysmeister replied:
There's got to be a better way of handling this without changing exiting behaviour and breaking existing playbooks. This is one of the oldest modules so it's probably used a fair bit more than the others.

cognifloyd replied:
OK. It turns out that it really doesn't matter if this section is unauthorized even if localhost exception doesn't apply, because the unauthorized error will be raised when adding the user. So, I backed out this part.

cognifloyd commented:
I added a simple integration test (and moved some tasks to separate include files so I could reuse them).
Plus, I dropped the behavioral changes I had in here before as they didn't really add any value.

@rhysmeister what do you think? Can you think of a more elegant way to do this while minimizing the required playbook tasks (ie simplifying the UX when the localhost exception applies)?

rhysmeister commented on tests/integration/targets/mongodb_user/tasks/main.yml:
```diff
@@ -221,3 +182,61 @@
+    that:
+    - mongodb_admin_user_created is changed
+
+- name: Check for presence of the create_for_localhost_exception file
```
Yep, exactly what I had in mind.
